### DISCUSSION TITLE
fluid zoom levels for web map ui

### DISCUSSION THREAD

**Discussion by u-an-i** [albar965/littlenavmap#665] [Ideas]
I believe https://api.kde.org/marble/html/classMarble_1_1MarbleWidget.html#a714c08d1ad92e216e77fcdc89190c3a0 (setDistance) is used for zoom parameter (named "distance") in webserver map commands. It appears to offer "arbitrary" precision for zoom.

Currently, webserver appears to ignore values between 2^n and 2^(n+1) for any n >0, falling back to the zoom which 2^n provides.

Additionally, there is https://api.kde.org/marble/html/classMarble_1_1MarbleWidget.html#ab3321f9d9fd1c48ffec36557bf66e1bb (setZoom). But it "is an abstract value without physical interpretation". From experiments, the distance parameter above might be fed into setZoom (it clamps at around 2^10 showing the whole globe which equals around 1000 which as zoom would do this according to the documentation).

setDistance appears to be fitting for zoming. Unfortunately it is not accompanied by a field of view: ie at a given distance, what is the angle area intersecting the earth surface producing a map boundary (which would be circular)? Think a cone intersecting a sphere where the cone can have any "opening angle".

**Comment by albar965** [comment]
LNM uses distance to zoom  everywhere. To avoid fuzzy maps (Marble bug) it zooms out one step after setting distance. So, you will not get what you set. Can be disabled in options. Not 100 percent sure if this also applies to the web UI.

This is quite a mess in Marble and required quite a few workarounds. Marble is also very crash-happy if you pass in wrong values like a zoom rectangle with width = 1.

About FoV: No idea how to solve this. Definitely not in the spherical projection. Looking at a pole you get a rectangle circumventing the whole globe, for example. Maybe just ignore the FoV issues for now.

I also tend to limit the web UI to Mercator projection.

**Reply by u-an-i** [reply]
I found that :) That zooIn(); zoomOut() of yours is what causes the fixed, non-continuous zoom i guessed. But i left this for last to clear up. First 1 minor actual logic change by passing distance from params rather than session when not zooming. The remaining code changes i already did i wanted to ask you about it first yesterday evening but then went on to do them in individual commits hopefully allowing "cherry-picking" or similar if you don't like them. (things like having the primary code flow in the if-case rather than the exception and return from that rather than jumping to else cases and back out to then return something -- negligible likely but i wanted to give the webserver any theoretical boost which is "obvious" (no micro "optmisation"), and also using switch() wehere possible rather than "lot"s of if..else -- just saw on CI it didn't compile, i need to dust off my c(++))

**Reply by u-an-i** [reply]
if you face troubles with Marble , how big do you reckon is the level of exertion to exchange it with an other solution?

I searched the web for Qt map solutions and found 3:

- https://www.mapbox.com/qt/
- https://developers.arcgis.com/qt/
- https://doc.qt.io/qt-5/qtlocation-index.html

The first 2 cost, partly per usage. Can that be manageable? If you contact their sales department and describe your project, perhaps they offer you a reduced rate when you add an "ad" :) ?

I looked at the Marble git and 1 commit message from last year stated preparational changes for anticipated Qt 6 changes.
If updating Marble or getting the current implemenation compatible is exhausting effort, exchange of Marble is worthwile?

I offer my support hereby, as good as I can (but this is not promised 100% of my time), unforeseen influences permitting.

**Reply by albar965** [reply]
See #180 Might take quite a few months. Therefore, I try to stick to it as long as possible. Qt 6 is not thaaaat important.
Only open source and free solutions.

Have some past experience at my job with Arcgis. Definitely not this.

I'd rather do some harsh changes to the Marble library. Remove application, useless astro lib, GPS, Qt Script references and more to lean it.

**Reply by u-an-i** [reply]
i understand,

do you mean opensource && costfree  or  opensource || costfree? Because mapbox offers a free tier (of which i don't know how this fits with your userbase) and their website looks smooth and clean. They offer raster, rastertile, vector, vectortile which appear interesting.

i was only downloading Qt 6 because I didn't know better. I guess I'm > 50 % through clearing errors for atools. But i postpone continuing this for now as completion and then perhaps lnm and marble will take  alot of time. On the other hand, perhaps Marble in its current state  (form KDE) works better, so it might be worth a try. But i don't know.

**Reply by albar965** [reply]
It's gotta combine with GPL.
Cost free _and_ opensource. Or cost free _for_ opensource like Qt.
[GDAL](https://gdal.org/https://gdal.org/) looks like the most professional solution so far.

But don't worry. This is far, far future.

**Reply by KOKAProduktion** [reply]
Hiho guys,

at the risk of getting on your nerves with OpenLayers :) Is there a critical reason why a web view based solution may not be suitable for the Qt in-app display?

In my imagination OL does the same thing as Marble but in JS. Is the Qt WebView too restricted for something like this?

I've added some thoughts about handling #180 at https://github.com/albar965/littlenavmap/discussions/677#discussioncomment-847070

Cheers

**Reply by albar965** [reply]
> Is the Qt WebView too restricted for something like this?

Isn't that based on Chrome? Would not expect any restrictions.

Didn't look at openlayers in detail. Does it come with projections? Spherical?

What about coordinate transformations from screen to world? This is performance critical.

On one side I'd like to get rid of Marble and on the other side I fear the effort. I tend to extend and fix the currently used LNM Marble branch.

**Reply by KOKAProduktion** [reply]
> Isn't that based on Chrome? Would not expect any restrictions.

Yes the QtWebEngine is based on Chromium. [qtwebengine-overview](https://doc.qt.io/qt-5/qtwebengine-overview.html)
It doesn't sound like trouble but I do have some experience embedding JS in WebViews in native applications, mimicking native behaviour and ever since I don't trust anything what i haven't tried out myself before. There is always some caveat, emerging in practice only, like for example limited access to console output, SSL-related issues as keychain access may be inexpectedly constrained, weird local fs access, customized/extended JS API etc.

For now I'd just need to know if you guys object to a WebView solution in general for some reason?

> Didn't look at openlayers in detail. Does it come with projections? Spherical?

OL utilizes [proj4](https://github.com/proj4js/proj4js) for standard and custom projection definitions. I can't tell yet if it provides all projections which Marble is providing. I'm quite sure it can't map in 3D space.

Related: 
https://openlayers.org/en/latest/doc/faq.html#what-projection-is-openlayers-using-

https://openlayers.org/en/latest/examples/reprojection-wgs84.html
https://openlayers.org/en/latest/examples/reprojection-image.html
https://openlayers.org/en/latest/examples/sphere-mollweide.html

> What about coordinate transformations from screen to world? This is performance critical.

You mean like "getLonLatByScreenPos" or "getScreenPosByLonLat"? Yes, this is included, in respect of the projection in use. However I'm in doubt if there is any Z (altitude) handling available out-of-the-box... I need to check that.

> On one side I'd like to get rid of Marble and on the other side I fear the effort. I tend to extend and fix the currently used LNM Marble branch.

I totally feel you. I can't promise anything at this time, but hopefully it becomes more evident what's possible or not when we get OL running as an [Web UI client implementation](https://github.com/albar965/littlenavmap/discussions/677#discussioncomment-847064). So I'd suggest postponing this kind of decisions until then.

I hope i got everything right...

Cheers!
